security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,391 @@
+	// function every time asked to present certificates to the server when a new connection is
+	// established. This is known as peer certificate reloading.
+	GetClientCertificate func(*tls.CertificateRequestInfo) (*tls.Certificate, error)
+	// VerifyPeer is a custom server authorization checking after certificate signature check.
+	// If this is set, we will replace the hostname check with this customized authorization check.
+	// If this is nil, we fall back to typical hostname check.
+	VerifyPeer CustomVerificationFunc
+	// ServerNameOverride is for testing only. If set to a non-empty string,
+	// it will override the virtual host name of authority (e.g. :authority header field) in requests.
+	ServerNameOverride string
+	RootCertificateOptions
+}
+
+// ServerOptions contains all the fields and functions needed to be filled by the client.
+type ServerOptions struct {
+	// ---------------General Rules For Certificate Setting-----------------------------------------
+	// Certificates or GetClientCertificate indicates the certificates sent from the server to
+	// the client to prove server's identities. The rules for setting these two fields are:
+	// Either Certificates or GetCertificate must be set; the other will be ignored
+	//----------------------------------------------------------------------------------------------
+	// If field Certificates is set, field GetClientCertificate will be ignored. The server will use
+	// Certificates every time when asked for a certificate, without performing certificate reloading.
+	Certificates []tls.Certificate
+	// If GetClientCertificate is set and Certificates is nil, the server will invoke this
+	// function every time asked to present certificates to the client when a new connection is
+	// established. This is known as peer certificate reloading.
+	GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	// ---------------General Rules For Root Certificate Setting------------------------------------
+	// RootCACerts or GetRootCAs indicates the the certificates trusted by the server side. The rules
+	// for setting these two fields are:
+	// If requiring mutual authentication on server side:
+	//     Either RootCACerts or GetRootCAs must be set; the other will be ignored
+	// Otherwise:
+	//     Nothing needed(the two fields will be ignored)
+	//----------------------------------------------------------------------------------------------
+	// If field RootCACerts is set, field GetRootCAs will be ignored. The server will use RootCACerts
+	// every time when verifying the peer certificates, without performing root certificate reloading.
+	RootCACerts *x509.CertPool
+	// If GetRootCAs is set and RootCACerts is nil, the server will invoke this function every time
+	// asked to check certificates sent from the client when a new connection is established.
+	// This is known as root CA certificate reloading.
+	GetRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
```

Comment:
Can we also use `RootCertificateOptions` instead of `RootCACerts` and `GetRootCAs`?